pkg/controller/devicetainteviction/device_taint_eviction.go

```diff
@@ -319,7 +320,8 @@ func New(c clientset.Interface, podInformer coreinformers.PodInformer, claimInfo
 }
 
 // Run starts the controller which will run until the context is done.
-func (tc *Controller) Run(ctx context.Context) {
+// An error is returned for startup problems.
+func (tc *Controller) Run(ctx context.Context) error {
```

Comment:
This is breaking the general approach that `Run` is not returning errors. None of our controller's `Run` does so. In your case you're adding event handlers in `Run` but that should happen in `New`. Although looking at the controller this will require a major rewrite :disappointed:

Reply:
Other controllers also fail in Run. They just hide it in their own code, some with error logging, some without:

https://github.com/kubernetes/kubernetes/blob/03e818744a8f92eb87155f2576dad982da7e6a43/pkg/controller/tainteviction/taint_eviction.go#L298-L301

https://github.com/kubernetes/kubernetes/blob/03e818744a8f92eb87155f2576dad982da7e6a43/pkg/controller/storageversiongc/gc_controller.go#L103-L106

Admittedly the setup procedure in this case in Run has more error paths (most of them very unlikely!). But in return you can call New and discard the result without any resource leaks (goroutines, registered informers). That is not true for the other controllers. Let's discuss more face-to-face...